[EXPERIMENTAL] NATS protocol parser

mtmk opened the pull request:
This is an experimental NATS Protocol Parser with zero allocations. Somewhat inspired by the .NET Runtime System.Text.Json library's Utf8JsonReader.

| Method | Mean     | Error    | StdDev   | Allocated |
|------- |---------:|---------:|---------:|----------:|
| Parse  | 862.5 ns | 320.3 ns | 17.56 ns |         - |

watfordsuzy commented on src/NATS.Client/JetStream/JetStream.cs:
```diff
@@ -0,0 +1,245 @@
+            return mh;
+        }
+
+        private PublishAck PublishSync(string subject, byte[] data, PublishOptions opts)
+            => PublishSync(new Msg(subject, null, MergeHeaders(null, opts), data), opts);
```
This will end up calling `MergeHeaders` twice if `opts` is not null and has headers (via `PublishSync(Msg, PublishOptions)`). Probably best to just use `null` for headers here if it is being passed into `PublishSync` that will merge them.

ColinSullivan1 replied:
Oh.. good catch.  Thanks!

watfordsuzy commented on src/NATS.Client/JetStream/JetStream.cs:
```diff
@@ -0,0 +1,245 @@
+        internal JetStream(IConnection connection, JetStreamOptions options)
+        {
+            Connection = connection;
+            Options = (options != null) ? options : JetStreamOptions.Builder().Build();
```
```suggestion
            Options = options ?? JetStreamOptions.Builder().Build();
```

ColinSullivan1 replied:
Thanks!  Been looking at too much java lately.  :)  I have a lot of changes in my branch so will make this change locally and push a bit later.

scottf replied:
Could also have been my code. I am trying to use do C# correctly.

watfordsuzy commented on src/NATS.Client/JetStream/JetStream.cs:
```diff
@@ -0,0 +1,245 @@
+            if (opts == null)
+                return null;
+
+            MsgHeader mh = headers != null ? headers : new MsgHeader();
```
```suggestion
            MsgHeader mh = headers ?? new MsgHeader();
```

ColinSullivan1 replied:
Will fix...

watfordsuzy commented on src/NATS.Client/JetStream/JetStream.cs:
```diff
@@ -0,0 +1,245 @@
+
+            if (!string.IsNullOrEmpty(opts.ExpectedLastMsgId))
+            {
+                if (mh == null) mh = new MsgHeader();
```
`mh` is always non-null here (and the other checks) due to line 78 always setting it to a non-null value. These `null` checks should be elided.

ColinSullivan1 replied:
Will fix - thanks.

scottf replied:
fixed

watfordsuzy commented on src/NATS.Client/JetStream/JetStreamMsg.cs:
```diff
@@ -0,0 +1,242 @@
+                    i++;
+                }
+            }
+            return i == 7;
```
How would this be 7? Seems to start at 8 and only ever go up.

ColinSullivan1 replied:
Not sure what I was originally thinking here.  Will fix...

scottf replied:
Look at JsPrefixManager, it manages other prefixes as well Also I think the default is `$JS.API.`  The Java client calls JsPrefixManager.HasPrefix with the reply to to determine if it's a JS message

scottf replied:
replaced by JsPrefixManager

watfordsuzy commented on src/NATS.Client/JetStream/JetStreamMsg.cs:
```diff
@@ -0,0 +1,242 @@
+
+        internal JetStream CheckReply(out bool isPullMode)
+        {
+            // TODO: Implement;
+            if ((sub is JetStreamSubscription) == false)
+            {
+                // Not a Jetstream enabled connection.
+                isPullMode = false;
+                return null;
+            }
+
+            isPullMode = (sub is JetStreamPullSubscription);
+            return ((JetStreamSubscription)sub).JS;
```
Could be simplified/.NETified some:
```suggestion
            if (sub is JetStreamSubscription jsSub)
            {
                isPullMode = jsSub is JetStreamPullSubscription;
                return jsSub.JS;
            }

            // Not a Jetstream enabled connection
            isPullMode = false;
            return null;
```

ColinSullivan1 replied:
+1

watfordsuzy commented on src/NATS.Client/JetStream/JetStreamMsg.cs:
```diff
@@ -0,0 +1,242 @@
+        /// <summary>
+        /// Gets the metadata associated with a JetStream message.
+        /// </summary>
+        public override MetaData Metadata { get; }
```
If the class is `MetaData` then I suggest the property matches:
```suggestion
        public override MetaData MetaData { get; }
```

scottf replied:
fixed

watfordsuzy commented on src/NATS.Client/JetStream/JetStreamMsg.cs:
```diff
@@ -0,0 +1,242 @@
+        internal MetaData(string metaData)
+        {
+            string[] tokens = metaData?.Split('.');
+            if (tokens.Length != 8)
```
`tokens` will be null if `metaData` is null:
```suggestion
            if (tokens?.Length != 8)
```

ColinSullivan1 replied:
Will fix...

watfordsuzy commented on src/NATS.Client/JetStream/PushSubscribeOptions.cs:
```diff
@@ -40,10 +27,17 @@ public sealed class PushSubscribeOptions
         {
-            _stream = stream;
-            _consumerConfiguration = consumerConfiguration;
+            string d = Validator.ValidateDurable(durable, false);
+            if (d == null && consumerConfiguration != null)
+            {
+                consumerConfiguration.Durable = Validator.ValidateDurable(consumerConfiguration.Durable, false);
+            }
+            consumerConfiguration.DeliverSubject = deliverSubject;
```
This is odd in a constructor where the only usage of `consumerConfiguration` is thru a base class.

scottf replied:
This has been changed. All validation is done in the build since the constructor is private. Line 40 here also can NPE as one of the tests showed.

watfordsuzy commented on src/NATS.Client/Msg.cs:
```diff
@@ -719,5 +741,68 @@ public bool HasHeaders
+        /// <summary>
+        /// Gets the metadata associated with a JetStream message.
+        /// </summary>
+        public virtual JetStream.MetaData Metadata { get { return null; } }
```
Same comment regarding the name of the property vs the class name.

watfordsuzy commented on src/NATS.Client/Msg.cs:
```diff
@@ -450,7 +472,7 @@ public IEnumerator GetEnumerator()
     /// received by the client application.
     /// </summary>
-    public sealed class Msg
+    public class Msg
```
Perhaps instead of un-sealing message you could add the `IsJetStream` property plus have the `JetStreamMsg` class implement a new `IJetStreamMsg` interface that contains the methods you'd need. In JetStream code you'd be able to do:
```csharp
if (msg is IJetStreamMsg jsMsg)
{
    msg.Ack();
}
```

scottf replied:
There is an `IsJetStream` property now and the connection instantiates a `Msg` or a `JetStreamMsg` determined by using the `JsPrefixManager`. Do you still think we need the `IJetStreamMsg` interface since the property is there? Seems like we'd prefer to use a built in method instead of checking a class type, but that may be Java vs .NET so looking for more guidance here.

AndyPook commented:
lots of places where _I_ would use `var`
is explicit typing a choice in this proj?

watfordsuzy commented:
My preference is the same as the .NET repo (sans Roslyn). Explicit types in all places except where obvious on the RHS (or the type is so terrible on the LHS that `var` improves readability).

[This project does not have a set-in-stone format and would not be considered idiomatic of modern .NET codebases.]

watfordsuzy commented on src/NATS.Client/JetStream/ConsumerConfiguration.cs:
```diff
@@ -118,7 +126,7 @@ public sealed class ConsumerConfigurationBuilder
             private string _deliverSubject;
             private long _startSeq;
-            private DateTime _startTime;
+            private DateTime _startTime = DateTime.MinValue; 
```
Why not a nullable DateTime?

watfordsuzy commented on src/NATS.Client/JetStream/JetStream.cs:
```diff
@@ -75,34 +76,34 @@ private MsgHeader MergeHeaders(MsgHeader headers, PublishOptions opts)
-                if (mh == null) mh = new MsgHeader();
+                mh = headers ?? new MsgHeader();
                 mh.Set(MsgHeader.ExpLastSeqHeader, opts.ExpectedLastSeq.ToString());
             }
```
This pattern of `mh = headers ?? new MsgHeader();` is not actually correct as used here. You would need to do:
```csharp
            MsgHeader mh = headers;

            if (!string.IsNullOrEmpty(opts.ExpectedLastMsgId))
            {
                mh = mh ?? new MsgHeader();
                // mh ??= new MsgHeader(); // C# 8.0
                mh.Set(MsgHeader.ExpLastIdHeader, opts.ExpectedLastMsgId);
            }
```
Otherwise, if `headers` is null then each of these blocks is going to re-assign `mh`.

scottf replied:
fixed as suggested

watfordsuzy commented on src/NATS.Client/JetStream/StreamConfiguration.cs:
```diff
@@ -430,6 +435,11 @@ public StreamConfigurationBuilder(StreamConfiguration sc)
         /// </summary>
         /// <returns>The StreamConfiguration</returns>
         public StreamConfiguration Build() {
+            if (_subjects == null)
```
```suggestion
            if (_subjects == null || !_subjects.Any())
```

scottf replied:
I went with this `if (_subjects == null || _subjects.Count < 1)` I wasn't sure is _subjects.Any() would handle null

AndyPook reviewed (commented):
> My preference is the same as the .NET repo (sans Roslyn). Explicit types in all places except where obvious on the RHS (or the type is so terrible on the LHS that `var` improves readability).
> 
> [This project does not have a set-in-stone format and would not be considered idiomatic of modern .NET codebases.]

_lots_ of places in dotnet repos where RHS is completely not obvious, but still uses `var` :)
lots of places in this repo like `List<string> results = new List<string>(values);` which is pretty "obvious". And lots of `var x = BuildThing()` which isn't

This is only the level of trivia (ie not important) and still "readable" so <shrug>
Just another "naming things" type discussion. Moving on...

scottf commented on src/NATS.Client/Internals/SimpleJson.cs:
```diff
@@ -1012,6 +1012,10 @@ internal override void WriteToStringBuilder(StringBuilder aSB, int aIndent, int
             foreach (var k in m_Dict)
             {
+                // Special NATS case: ignore null values or negative numbers.
+                if (k.Value.Value == null || k.Value.AsInt < 0)
```
Why are we ignoring negative numbers? Changing k.Value.AsInt to k.Value.AsLong

ColinSullivan1 replied:
This was a bug... thanks.

scottf commented on src/NATS.Client/JetStream/ConsumerConfiguration.cs:
```diff
@@ -113,12 +121,12 @@ public static ConsumerConfigurationBuilder Builder(ConsumerConfiguration cc)
             private string _deliverSubject;
             private long _startSeq;
-            private DateTime _startTime;
+            private DateTime _startTime = DateTime.MinValue; 
```
I think it's ok to leave this. It's not obvious that it's a value type as a search on the internet shows that this question is asked a lot. Plus it's explicit as to what the default is.

AndyPook replied:
it is "ok", but a bit confusing to a dotnet geek :)
kinda the same as explicitly setting the `long` on the previous line to `0`. Or the `string`s to `null`.

Again, not "important", just looks a bit "odd"

scottf replied:
I definitely agree where privatives are concerned. Can other dotnet devs please chime in regarding DateTime?

sixlettervariables replied:
We would omit the MinValue assignment. 

We'd also prefer nullable if optional instead of relying on the default value which is less obvious.

scottf commented on src/NATS.Client/JetStream/JetStreamMsg.cs:
```diff
@@ -0,0 +1,245 @@
+    {
+        static readonly private byte[] ackAck = Encoding.ASCII.GetBytes("+ACK");
+	    static readonly private byte[] ackNak = Encoding.ASCII.GetBytes("-NAK");
+	    static readonly private byte[] ackProgress = Encoding.ASCII.GetBytes("+WPI");
```
All are developers converting tabs to spaces?

scottf commented on src/NATS.Client/JetStream/JetStreamMsg.cs:
```diff
@@ -0,0 +1,245 @@
+                replySubj[2] != 'S' ||
+                replySubj[3] != '.' ||
+                replySubj[4] != 'A' ||
+                replySubj[5] != 'C' ||
```
This is being replaced as it does not handle custom prefixes.

scottf commented on src/NATS.Client/JetStream/JetStreamMsg.cs:
```diff
@@ -0,0 +1,245 @@
+            }
+
+            int j = 0;
+            for (int i = 8; i < len; i++)
```
Being replaced...

scottf commented on src/NATS.Client/Msg.cs:
```diff
@@ -82,6 +82,26 @@ public sealed class MsgHeader : IEnumerable
+        /// <summary>
+        /// JetStream expected message ID header name.
+        /// </summary>
+        public static readonly string MsgIdHeader = "Nats-Msg-Id";
```
changing to `public const`

scottf commented on src/NATS.Client/JetStream/ApiResponse.cs:
```diff
@@ -27,6 +27,14 @@ public abstract class ApiResponse
 
         internal ApiResponse(Msg msg) : this(Encoding.UTF8.GetString(msg.Data)) { }
```
so I built this, I'm a Java developer, so there may be some of that at work, sub classing and all... but this actually should never be called with a null message, so leaving it un guarded will suss out coding problems.

AndyPook replied:
it's a style choice I guess
The current way means that _every_ sub-class must do the null check. You just need one to forget.

"leaving it un guarded will suss out coding problems" ummm, ok. This is what guards are normally for. So that you get an meaningful exception in a useful place, rather than down in the guts of dotnet somewhere and then need to study the stack.

I thought java generally had the same style?

anyway, not super important :)

scottf replied:
So now that I think of it, this would never be called with a null message.  There is no need for a code guard here. These constructors are internal and it's expected that the developer know what they are doing. I would use a code guard when a valid value is null, but null is not valid here. On some languages, there are ways to mark parameters as not nullable. Is there the equivalent in .NET, I would use that I suppose

sixlettervariables replied:
Nullable reference types are available in the latest language version yes.

ColinSullivan1 replied:
> lots of places where _I_ would use `var`
> is explicit typing a choice in this proj?

It was simply a preference in initial development but I have no aversion to using var moving forward when it makes sense.

https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#implicitly-typed-local-variables

sergeyshaykhullin commented on src/Tests/UnitTests/TestMessageHeaders.cs:
```diff
@@ -277,6 +285,76 @@ public void TestHeaderExceptions()
+        }
+
+        private void shouldNATSInvalidHeaderException(string s)
+        {
```
Should this be an ASCII?

scottf replied:
sure I suppose it's just a test though. I'll fix anyway

watfordsuzy commented on src/NATS.Client/JetStream/Error.cs:
```diff
@@ -33,18 +34,25 @@ internal Error(JSONNode node)
-                return Code == NOT_SET
-                    ? "Unknown JetStream Error: " + _node.ToString()
-                    : "Unknown JetStream Error (" + Code + ")";
+            if (Code == NOT_SET) {
+                if (ApiErrorCode == NOT_SET) {
```
```suggestion
            if (Code == NOT_SET)
            {
                if (ApiErrorCode == NOT_SET)
                {
```

scottf replied:
will fix

watfordsuzy commented on src/NATS.Client/JetStream/Error.cs:
```diff
@@ -33,18 +34,25 @@ internal Error(JSONNode node)
             }
 
-            return Desc + " (" + Code + ")";
+            if (ApiErrorCode == NOT_SET) {
+                return Desc + " (" + Code + ")";
```
```suggestion
            if (ApiErrorCode == NOT_SET)
            {
                return $"{Desc} ({Code})";
```

scottf replied:
will fix

watfordsuzy commented on src/NATS.Client/JetStream/Error.cs:
```diff
@@ -33,18 +34,25 @@ internal Error(JSONNode node)
+                return Desc + " (" + Code + ")";
+            }
+
+            return Desc + " (" + Code + ") [" + ApiErrorCode + "]";
```
```suggestion
            return $"{Desc} ({Code}) [{ApiErrorCode}]";
```

scottf replied:
will fix

watfordsuzy commented on src/NATS.Client/JetStream/Error.cs:
```diff
@@ -33,18 +34,25 @@ internal Error(JSONNode node)
+                if (ApiErrorCode == NOT_SET) {
+                    return Desc;
+                }
+                return Desc + " [" + ApiErrorCode + "]";
```
```suggestion
                return $"{Desc} [{ApiErrorCode}]";
```

scottf replied:
Just force of habit on the brackets, I'll fix even though it's clearly not as good! :) Also I will use the interpolation, that's much better than the concatenation

watfordsuzy replied:
I do a lot of open source Java on the side and I rely heavily on VS Code's reformatting to make it more Java like...so I definitely get it!

watfordsuzy reviewed (commented):
Small stuff, places with Java style braces should be reformatted but I only tagged the first one. Perhaps since this is a port of the Java client we can wait for a bit and reformat it all once the bulk of it is in place.

watfordsuzy commented on src/NATS.Client/Internals/JsonUtils.cs:
```diff
@@ -73,5 +74,10 @@ internal static JSONArray ToArray(List<string> list)
+ 
+        public static byte[] SimpleMessageBody(string name, object value)
+        {
+            return Encoding.ASCII.GetBytes("{\"" + name + "\":" + value.ToString() + "}");
```
Also, will we need to consider escaping of `value` here, or is that on the caller?
```suggestion
            return Encoding.UTF8.GetBytes("{\"" + name + "\":" + value.ToString() + "}");
```

scottf replied:
right now the only use is where the object is a number. Also internal uses should never have anything other than ascii

watfordsuzy replied:
Suggested as JSON is UTF8 encoded. I didn't suggest interpolation due to the braces, otherwise the rules for escaping quotes are based on whether or not the string is literal or not (i.e. prefixed with `@`):
```suggestion
            return Encoding.UTF8.GetBytes($"{{\"{name}\":{value}}}");
```
An alternative using an interpolated literal string:
```csharp
            return Encoding.UTF8.GetBytes($@"{{""{name}"":{value}}}");
```

watfordsuzy reviewed (commented):
One note about JSON encoding.

sixlettervariables commented on src/NATS.Client/Internals/JsonUtils.cs:
```diff
@@ -75,9 +75,14 @@ internal static JSONArray ToArray(List<string> list)
         {
-            return Encoding.ASCII.GetBytes("{\"" + name + "\":" + value.ToString() + "}");
+            return Encoding.ASCII.GetBytes("{\"" + name + "\":" + value + "}");
+        }
```
We need to be careful marshalling `long` to and from JSON as we only basically have 53 bits.

scottf replied:
Can you clarify what you mean? 53 or 63 bits? SimpleMessageBody is just a convenience method. Maybe it should be `internal`? Is ToString redundant? The current use is for a message sequence number, which in go is a uint64. I'll try to provide an overload.

sixlettervariables replied:
JSON cannot marshal integer values beyond 53-bits because JavaScript lacks an integer type all together. We need to send these as strings if they are in fact UInt64.

Examples of problems:
- https://randomascii.wordpress.com/2020/09/27/floating-point-in-the-browser-part-1-impossible-expectations/
- https://www.kdab.com/fixing-bugs-via-lateral-thinking/

scottf replied:
Our schema expects a number. I'll verify that the .NET client handles it fine. I don't think it's our problem if javascript in browsers can't handle it. Also 53 bits is a pretty huge number so I think we'll be okay. The java client currently only supports 32 bits (signed long) and it we have not heard any issues. I'm fixing that anyway though.

watfordsuzy commented on src/NATS.Client/JetStream/IJetStreamSubscription.cs:
```diff
@@ -93,6 +93,6 @@ public interface IJetStreamPullSubscription : IJetStreamSubscription, ISyncSubsc
         /// <param name="maxWaitMillis">The maximum time to wait for the first message.</param>
         /// <returns></returns>
-        JetStreamMsg[] Fetch(int batchSize, long maxWaitMillis);
+        List<Msg> Fetch(int batchSize, int maxWaitMillis);
```
```suggestion
        IList<Msg> Fetch(int batchSize, int maxWaitMillis);
```

scottf replied:
got it. List is concrete like ArrayList in java, IList is the interface like List is in Java. Fixed

watfordsuzy commented on src/NATS.Client/JetStream/JetStream.cs:
```diff
@@ -88,6 +89,24 @@ private PublishAck PublishSyncInternal(string subject, byte[] data, MsgHeader hd
+                return null;
+            }
+
+            Duration timeout = options == null ? JetStreamOptions.RequestTimeout : options.StreamTimeout;
```
We'll want a way to do infinite timeout, but with the `timeout.Millis` call below performing math I'm not sure that's as simple anymore.

scottf replied:
The timeout applies to the publish portion and is required.

watfordsuzy commented on src/NATS.Client/JetStream/JetStream.cs:
```diff
@@ -88,6 +89,24 @@ private PublishAck PublishSyncInternal(string subject, byte[] data, MsgHeader hd
+            Duration timeout = options == null ? JetStreamOptions.RequestTimeout : options.StreamTimeout;
+
+            return Conn.RequestAsync(msg, timeout.Millis)
+                .ContinueWith(async antecedent => ProcessPublishResponse(antecedent.Result, options), CancellationToken.None)
```
Drop the `async` if you are not using `await`:
```suggestion
                .ContinueWith(antecedent => ProcessPublishResponse(antecedent.Result, options), CancellationToken.None)
```

watfordsuzy replied:
On second thought, is there a reason that `PublishAsyncInternal` cannot be declared `async` and you `await Conn.RequestAsync(...)`?

scottf replied:
Conn.RequestAsync already returns a future. It hides the async under the covers. I understood that antecedent just tacks itself on to the task.

watfordsuzy replied:
In general you don't really want to use ContinueWith if it can be avoided. Lots of little caveats here and there as to how the task is attached, cancellation, exceptions; best to use `async` + `await` if you're not just immediately chaining a result. It can also get you into trouble if at any point you mix sync-over-async and async code.

scottf replied:
Adding an issue so we don't lose track of this. https://github.com/nats-io/nats.net/issues/475

mtmk committed:
NATS protocol parser

This is an experimental NATS Protocol Parser with zero allocations. Somewhat
inspired by the .NET Runtime System.Text.Json library's Utf8JsonReader.

| Method | Mean     | Error    | StdDev   | Allocated |
|------- |---------:|---------:|---------:|----------:|
| Parse  | 862.5 ns | 320.3 ns | 17.56 ns |         - |

mtmk committed:
wip

mtmk commented:
too old. need a fresh look at the solution in a clean way.
